### DISCUSSION TITLE
How do you deal with project-specific settings (e.g. formatting)

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [LazyVim/LazyVim#1878] [Q&A] [answered]
How do you deal with going from project to project where e.g. different formatters are expected?

Let's for example say you are jumping between different Python git projects. In one, the `black` formatter is expected and in another, the `yapf` formatter is excpected. When formatting on save, you'd have to e.g. hardcode local project paths into a lua config that modifies the behavior/setup of `conform.nvim`, or perhaps you could add some sort of file into the git project itself which would get read.

I would personally prefer the latter, as I keep my configs in my public dotfiles repo, and I wouldn't want to expose private git project names/paths there. I'm just not really sure how to pull this off for e.g. `conform.nvim` in a good way, preferably leveraging lazy.nvim… this is what I have, currently:

```lua
-- options.lua

vim.opt.exrc = true -- allow local .nvim.lua .vimrc .exrc files
vim.opt.secure = true -- disable shell and write commands in local .nvim.lua .vimrc .exrc files
```

…and then I place a `.nvim.lua` file in the project root folder from which I launch Neovim:
https://github.com/fredrikaverpil/dotfiles/blob/main/templates/.nvim.lua

But I wish I could add a lazy.nvim override in here, which overrides whatever I have in my lazyvim config. Doesn't seem like it's possible...?

Would be interesting to hear how you go about solving this problem.

**Comment by danielo515** [comment]
I'm also interested on this to set the root detection mechanism in those projects where git is the only reliable way

**Comment by dimas-cyriaco** [comment]
Maybe you can use [neoconf](https://github.com/folke/neoconf.nvim) for this.

**Reply by danielo515** [reply]
I thought about that, but neoconf only seems to allow lsp configuration values, not random neovim configurations

**Comment by jyuan0** [comment]
For conform.nvim, you could pass a function (`function(bufnr) ...`) for the formatter for a language.

**Comment by fredrikaverpil** [comment]
I really wish I could add some sort of lazy.nvim override in the project's `.nvim.lua`. Doesn't seem possible to me though.

**Reply by abeldekat** [reply]
You're question has been in the back of my mind for quite some time.
Today, I found a solution I would like to share.
You can require module `config.lazy` from your project's `.nvim.lua` when applying some minor [modifications](https://github.com/abeldekat/starter/commits/lazyvim_exrc).

Modify:

1. `config.lazy.lua` to return a function accepting an extra spec
2. `init.lua` to return without loading when a local `.nvim.lua` is present

`init.lua`:

```lua
if vim.loop.fs_stat(".nvim.lua") then
  vim.opt.exrc = true -- allow local .nvim.lua .vimrc .exrc files
  return
end

-- bootstrap lazy.nvim, LazyVim and your plugins
require("config.lazy")()
```

`config.lazy.lua`, see [branch lazy_exrc](https://github.com/abeldekat/starter/blob/bf0165cbbb564c686a1666833eaf7de4f0349a0d/lua/config/lazy.lua#L9):

```lua
--- code

return function(extraspec)
   local spec = {
      --- code, the spec
   }
   require("lazy").setup(
     --- code
    spec = extraspec and vim.list_extend(spec, extraspec) or spec,
     --- more code
   )
end
```

For a proof of concept:

`lazytest installation`:

```sh
cd ~/.config
git clone https://github.com/abeldekat/starter.git lazytest
cd lazytest
git checkout lazyvim_exrc
NVIM_APPNAME=lazytest nvim init.lua
# reboot after installation
```

`lazytest project`:
```sh
mkdir -p  ~/projects/lazytest
cd ~/projects/lazytest
touch .nvim.lua
git init
git add .
git commit -m "dummy commit"
```

`~/projects/lazytest/.nvim.lua`:

```lua
-- bootstrap lazy.nvim, LazyVim and your plugins
require("config.lazy")({
  {
    "telescope.nvim",
    keys = {
      {
        "<leader>z", -- proof the concept by adding a key to telescope
        function()
          require("lazyvim.util").telescope("live_grep")()
        end,
        desc = "Grep (root dir)",
      },
    },
  },
})
```

Result: Only when starting `NVIM_APPNAME=lazytest nvim` from within `~/projects/lazytest` the new telescope key `<leader> z` is available.

**Reply by dpetka2001** [reply]
Awesome work!! Really love this solution. Thanks a lot @abeldekat :smile:

**Comment by abeldekat** [comment, marked as answer]
Also see this PR: [feat(extras): Enable project-specific plugin specs using local .lazy.lua](https://github.com/LazyVim/LazyVim/pull/2115)

**Reply by fredrikaverpil** [reply]
@abeldekat this all looks super nice, I just haven't had time to look at it and try it out yet. Having this officially supported for sure would be super awesome!

**Reply by fredrikaverpil** [reply]
Awesome job on this @abeldekat ❤️  Thank you so, so much!

**Reply by abeldekat** [reply]
You're welcome!

**Comment by kobzar** [comment]
Maybe you can use some keybindings to use different formatters for?
like 
SPS fb = require("conform").formatters_by_ft.python = { "black" }
SPS fy = require("conform").formatters_by_ft.python = { "yapf" }

**Reply by danielo515** [reply]
That is a workaround for just one of the problems, which is formatting. Sometimes you want other configurations project specific, not only formatting

**Reply by fredrikaverpil** [reply]
@kobzar formatting was just an example. I would like to have per-project settings on certain things.

**Comment by yanshay** [comment]
This issue bugged me for a while until it was too painful...
So I developed a simple solution, project settings overrides expressed in JSON (more secure than lua).
https://gist.github.com/yanshay/43c0f8257f9dd25ac8e33359eac8d854
Disclaimer: I tested it only on my very simple use case, so not sure if it's general enough, but surely an interesting direction.

**Comment by rubiin** [comment]
I dont know if this is the correct place for this but this seems like vscode profiles. I was also looking for something similar but for language specific profiles

**Comment by chingDev0fficial** [comment]
when I open Lazyvim

nvim

and select the project by clicking the "p"

and selecting my project folder then after this I got redirected to init.lua

can someone help me please

**Comment by tobbbe** [comment]
I want to disable all formatting by default, and then set it up per project with .lazy.lua.

Is the way to do it?

1. Set in options.lua
```
vim.g.autoformat = false
vim.g.lazyvim_prettier_auto_format = false
vim.g.lazyvim_eslint_auto_format = false
vim.g.lazyvim_prettier_needs_config = true
``` 

2. Set in conform.lua
```
return {
  {
    "stevearc/conform.nvim",
    -- disable all formatters. Set up manually with .lazy.lua per project instead.
    opts = {
      formatters_by_ft = {
        javascript = {},
        typescript = {},
        javascriptreact = {},
        typescriptreact = {},
        css = {},
        scss = {},
        less = {},
        html = {},
        json = {},
        jsonc = {},
        yaml = {},
        markdown = {},
        graphql = {},
        handlebars = {},
        vue = {},
        svelte = {},
      },
      default_format_opts = {
        -- by default, conform uses lsp formatting as fallback. Disable that.
        -- want to disable per lsp? see https://github.com/LazyVim/LazyVim/discussions/2159
        lsp_format = "never",
      },
    },
  },
}
``` 

And then in each project Ill add a .lazy.lua with for ex:
```
vim.g.autoformat = true

return {
    {
        "stevearc/conform.nvim",
        opts = {
        formatters_by_ft = {
            javascript = { "prettier" },
            typescript = { "prettier" },
            javascriptreact = { "prettier" },
            typescriptreact = { "prettier" },
        },
        },
    },
}

```

**Reply by dragonfly8677** [reply]
Did you figure out a way to do this?

**Reply by tobbbe** [reply]
Yea! this works great

conform.lua
```lua
-- Disable all formatters by default. Also check options.lua

return {
  {
    "stevearc/conform.nvim",
    -- disable all formatters. Set up manually with .lazy.lua per project instead.
    opts = {
      formatters_by_ft = {
        javascript = {},
        typescript = {},
        javascriptreact = {},
        typescriptreact = {},
        css = { "prettierd" },
        scss = {},
        lua = { "stylua" },
        less = {},
        html = {},
        json = {},
        jsonc = {},
        yaml = {},
        markdown = {},
        graphql = {},
        handlebars = {},
        vue = {},
        svelte = {},
      },
      default_format_opts = {
        -- by default, conform uses lsp formatting as fallback. Disable that.
        -- want to disable per lsp? see https://github.com/LazyVim/LazyVim/discussions/2159
        lsp_format = "never",
      },
    },
  },
}
``` 

options.lua
```lua
vim.g.autoformat = false
vim.g.lazyvim_prettier_auto_format = false
vim.g.lazyvim_eslint_auto_format = false
vim.g.lazyvim_prettier_needs_config = true

``` 

And then i add a .lazy.lua in each project with ex:
```lua
vim.g.autoformat = true

return {
	{
		"stevearc/conform.nvim",
		opts = {
			formatters_by_ft = {
				javascript = { "prettierd" },
				typescript = { "prettierd" },
				javascriptreact = { "prettierd" },
				typescriptreact = { "prettierd" },
			},
		},
	},
}

-- or
vim.g.autoformat = true

return {
	{
		"stevearc/conform.nvim",
		opts = {
			formatters_by_ft = {
				javascript = { "eslint_d" },
				typescript = { "eslint_d" },
				javascriptreact = { "eslint_d" },
				typescriptreact = { "eslint_d" },
			},
		},
	},
}

```

**Reply by dragonfly8677** [reply]
Thank you!!
